pkg/itemoperation/backup_operation.go

```diff
@@ -0,0 +1,44 @@
+/*
+Copyright the Velero contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package itemoperation
+
+// BackupOperation stores information about an async item operation
+// started by a BackupItemAction plugin (v2 or later)
+type BackupOperation struct {
+	Spec BackupOperationSpec `json:"spec"`
+
+	Status OperationStatus `json:"status"`
+}
+
+type BackupOperationSpec struct {
+	// BackupName is the name of the Velero backup this item operation
+	// is associated with.
+	BackupName string `json:"backupName"`
+
+	// BackupUID is the UID of the Velero backup this item operation
+	// is associated with.
+	BackupUID string `json:"backupUID"`
+
+	// BackupItemAction is the name of the BackupItemAction plugin that started the operation
+	BackupItemAction string `json:"backupItemAction"`
+
+	// Kubernetes resource identifier for the item
+	ResourceIdentifier string "json:resourceIdentifier"
+
+	// OperationID returned by the BIA plugin
+	OperationID string "json:operationID"
+}
```

Comment:
*Also applied to RestoreOperation*
Except the plugin restart case, what else is the persisted json used for?

Reply:
It will also be used in `velero describe` to give user information on the operations associated with a backup/restore, and it may also be used on subsequent reconciles -- i.e. backup 1 is WaitingForPluginOperations, so velero executes backup 2. When it later goes back to reconcile backup 1, it will pull this file down and check status for any unfinished backup. We don't necessarily need to keep every unfinished backup/restore in memory until the longstanding operations are done when Velero isn't otherwise processing the backup/restore.

Reply:
Can I say that - the plugin restart doesn't use this json?
Previously, I thought the plugin requires the info in the json in order to recovery the connection with the backup, the operationID and the CR.
By your explanations and after further thinking, I think this is not required:

- Velero doesn't know the restart of the plugin server, so it calls the BIA V2 operation with the same OperationID and backup name
- A new plugin server is started by the plugin framework
- The plugin server now has the OperationID and the backup name
- From the backup name, it can find the CR, because when creating the CR we can add something like an annotation to the CR to indicate that when backup is the CR for

So the json is not used at all. Am I correct?

Reply:
>  i.e. backup 1 is WaitingForPluginOperations, so velero executes backup 2. When it later goes back to reconcile backup 1, it will pull this file down and check status for any unfinished backup
This idea sounds good. But let's think one more step:
In this case, after backup 1 goes to WaitingForPluginOperations and Velero starts to execute backup 2, most probably, the plugin client to backup 1 will be closed at this time. Then the plugin server will also close.
It means, this comes to a plugin restart case, it comes to the question - does the plugin recovery after restart must succeed? If the answer is no, then when Velero reconcile backup 1 later, it cannot call any BIA V2 operation at this 
time, because the call would return a failure.

Another question is, who is going to update the status in the file since Velero has started to execute backup 2?

Reply:
@Lyndon-Li That's right. The plugin itself never needs access to this json. Rather Velero's controllers need the json content in order to know which plugins to poll for progress updates. The plugin only ever needs information from Velero as defined by the plugin API -- i.e. for the `Execute` call, it uses the Execute input args, for the `Progress` call it uses the defined Progress input args.

Reply:
As far as I understand the way the plugin containers in practice, the plugin server only restarts when the container restarts. Ending a backup and starting a new one doesn't do anything to registered plugins. The plugins receive calls for each operation registered for each resource that matches. So the plugin will just receive a series of individual requests. So backup1 may send 100 plugin Execute calls. Then maybe 20 Progress calls for async operations. Then backup2 may send 200 Execute calls, followed by 10 Progress calls. Then we check backup1 again, so maybe there are only 15 Progress calls (since 5 of them had already completed in the first iteration), etc. There is no plugin-container-level event associated with finishing a backup.

Reply:
Velero will update the itemoperations.json status each time is runs through the list of outstanding operations and calls Progress on them for a given backup. So in the above example, backup1 completes the main backup process and initial round of Progress calls, updates itemoperations along with the rest of the backup (backup json, log, etc.). Then it runs through backup2. Then it reconciles backup1 again, making progress calls on the outstanding operations, and if there are any status changes, updates the itemoperations json file again via the object store API.

Reply:
> the plugin server only restarts when the container restarts
Let me further test this by debugging.

Reply:
> Then it reconciles backup1 again, making progress calls on the outstanding operations, and **if there are any status changes**, updates the itemoperations json file again via the object store API

Thanks for the explanation. 
As far as my understanding, frequently persisting the json will involve many network activities and further more, this may not be friendly to the future backup repository features. 
If the purpose is for saving memory only, I would suggest we keep the json for every backup in memory, this will not take too much memory because the in-parallel backup count is limited and a single json is with limited size. For example, let say each json is 10KB and we have 1000 in-parallel backups, the total memory usage is around 10 MB, which is totally acceptable.

Reply:
@Lyndon-Li That makes sense. It may be worth holding off on certain json updates, keeping things in memory for performance reasons. None of that affects this PR, though -- the interface for uploading/downloading this json if/when we do want it is unchanged. We will definitely upload it the first time we update the backup (after initial processing and first pass at operation progress). This will ensure that the full list of operations is persisted so that velero can recover from a restart after this point). If we decide *not* to update the backup store on every intermediate reconcile, then we can implement the in-memory cache you propose above, and on any subsequent reconcile (i.e. after initial backup run), we check for in-memory progress struct first. if found, we use it, else pull it from object store. At the end of reconcile, if we're completing the backup, we upload it and remove it from the in-memory list of current backups, and if the backup is still not done, we just update in-memory store and move on.

The two main consequences of the above:
1) we still preserve resilience to Velero restarts -- if a backup has updated operation progress that's not uploaded, on the first reconcile post-restart, Velero will have to make progress calls for operations that had previously been marked completed in-memory but not persisted -- that's OK, though. The operation will still be complete, so the result should be the same.
2) It's possible that `velero describe` will report out-of-date information. I'm not sure whether client-side describe calls will be able to access the in-velero-pod-memory progress data. One way around this would be to modify the DownloadRequest controller and in the case of DownloadRequests for a backup/restore's item operations, that controller could force updating the BSL json data. We can work through that in the context of the describe implemetation.

Reply:
This sounds good, thanks. For any other details, let's discuss in the implementation PR of backup controller and download request controller.